ADALiOS/ADALiOS/ADAuthenticationRequest.m

```diff
@@ -79,18 +77,146 @@ - (id)initWithContext:(ADAuthenticationContext*)context
     return self;
 }
 
-#define CHECK_REQUEST_STARTED { \
+#define CHECK_REQUEST_STARTED_R(_return) { \
     if (_requestStarted) { \
         NSString* _details = [NSString stringWithFormat:@"call to %s after the request started. call has no effect.", __PRETTY_FUNCTION__]; \
         AD_LOG_WARN(_details, nil); \
-        return; \
+        return _return; \
     } \
 }
 
-- (void)setScope:(NSString *)scope
+#define CHECK_REQUEST_STARTED CHECK_REQUEST_STARTED_R()
+
+static NSArray* _arrayOfLowercaseStrings(NSArray* strings, NSString* context, ADAuthenticationError* __autoreleasing * error)
+{
+    if (!strings || ![strings count])
+    {
+        ADAuthenticationError* adError = [ADAuthenticationError invalidArgumentError:@"%@ cannot be nil or empty", context];
+        if (error)
+        {
+            *error = adError;
+        }
+        return nil;
+        
+    }
+    NSMutableArray* lowercase = [[NSMutableArray alloc] initWithCapacity:[strings count]];
+    
+    for (NSString* string in strings)
+    {
+        if (![string isKindOfClass:[NSString class]])
+        {
+            ADAuthenticationError* adError = [ADAuthenticationError invalidArgumentError:@"%@ contains non-string objects.", context];
+            if (error)
+            {
+                *error = adError;
+            }
+            
+            return nil;
+        }
+        
+        [lowercase addObject:[string lowercaseString]];
+    }
+    
+    return lowercase;
+}
+
+static ADAuthenticationError* _validateScopes(NSArray* scopes)
+{
+    if ([scopes containsObject:@"openid"] || [scopes containsObject:@"offline_access"])
+    {
+        return [ADAuthenticationError invalidArgumentError:@"Can not pass in \"openid\" or \"offline_access\" scopes"];
+    }
+    
+    return nil;
+}
+
+//static BOOL isClientID(NSString* scope)
+//{
+    
+  //  NSError *error;
+    
+ //   NSRegularExpression *regex =
+ //   [NSRegularExpression regularExpressionWithPattern:@"\\A\\{[A-F0-9]{8}-[A-F0-9]{4}-[A-F0-9]{4}-[A-F0-9]{4}-[A-F0-9]{12}\\}\\Z"
+                                 //             options:NSRegularExpressionAnchorsMatchLines
+                                 //               error:&error];
+ //   NSPredicate *testGUID = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", regex];
+    
+    
+    
+ //   if ([testGUID evaluateWithObject: scope]) {
+        
+   //     return YES;
+        
+  //  } else {
+        
+  //      return NO;
+  //  }
+//}
+
+- (ADAuthenticationError*)setScopes:(NSArray *)scopes
+{
+    CHECK_REQUEST_STARTED_R(nil);
+    
+    ADAuthenticationError* error = nil;
+    NSMutableArray* lowercaseScopes = _arrayOfLowercaseStrings(scopes, @"scopes", &error);
+    if (!lowercaseScopes)
+    {
+        return error;
+    }
+    
+    RETURN_IF_NOT_NIL(_validateScopes(lowercaseScopes));
+    
+    for (NSString* scope in lowercaseScopes) {
+        
+        if ([scope isEqualToString:_clientId]) {
```

Comment:
We need to fail instead of altering user input

Sent from Outlook<http://aka.ms/Ox5hz3>




On Wed, Sep 16, 2015 at 4:07 PM -0700, "Brandon Werner" <notifications@github.com<mailto:notifications@github.com>> wrote:


In ADALiOS/ADALiOS/ADAuthenticationRequest.m<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzureAD%2fazure-activedirectory-library-for-objc%2fpull%2f378%23discussion_r39696354&data=01%7c01%7cKanishk.Panwar%40microsoft.com%7c153f04cc37324dea36b508d2beeba744%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=%2fEhsIj9BSQSyoBmuuqbnRnP97IfuLFRKiso3tcf%2bBFI%3d>:

> +- (ADAuthenticationError*)setScopes:(NSArray *)scopes
> +{
> +    CHECK_REQUEST_STARTED_R(nil);
> +
> +    ADAuthenticationError* error = nil;
> +    NSMutableArray* lowercaseScopes = _arrayOfLowercaseStrings(scopes, @"scopes", &error);
> +    if (!lowercaseScopes)
> +    {
> +        return error;
> +    }
> +
> +    RETURN_IF_NOT_NIL(_validateScopes(lowercaseScopes));
> +
> +    for (NSString* scope in lowercaseScopes) {
> +
> +        if ([scope isEqualToString:_clientId]) {


Like the highlander, there is now only one. I remove all scopes upon entering clientID flow.

   RETURN_IF_NOT_NIL(_validateScopes(lowercaseScopes));

    for (NSString* scope in lowercaseScopes) {

        if ([scope isEqualToString:_clientId]) {

            // first, remove all scopes as if Client ID is present there can only be the claims below.

            [lowercaseScopes removeAllObjects];

            // next, let's add the scopes that we need to get an id_token. These will be the only scopes.

            [lowercaseScopes addObject:@"openid"];
            [lowercaseScopes addObject:@"offline_access"];

        }
        ```


-
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzureAD%2fazure-activedirectory-library-for-objc%2fpull%2f378%2ffiles%23r39696354&data=01%7c01%7cKanishk.Panwar%40microsoft.com%7c153f04cc37324dea36b508d2beeba744%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=ufaJcvRIXCxS8EcQ2yGJSnd5qTAm15BAhY0I2cWhTtk%3d>.